text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,391 @@
+| 3                | 1/16                  | 1-in-8 and above       |
+| ...              | ...                   | ...                    |
+| 0 <= `r` <= 60   | 1/(2**(-`r`-1))       | 1-in-2**-`r` and above |
+| ...              | ...                   | ...                    |
+| 58               | 2**-59                | 1-in-2**-58 and above  |
+| 59               | 2**-60                | 1-in-2**-59 and above  |
+| 60               | 2**-61                | 1-in-2**-60 and above  |
+| 61               | 2**-61                | 1-in-2**-61 and above  |
+
+Such a random variable `r` can be generated using the following
+pseudocode.
+
+```golang
+func nextRandomness() int {
+  r := 0
+  for r < 61 && nextRandomBit() == false {
+    r++
+  }
+  return R
+}
+```
+
+This can be computed from a stream of random bits as the number of
+leading zeros using efficient instructions on modern computer
+architectures.
+
+For example, the value 3 means there were three leading zeros and
+corresponds with being sampled at probabilities 1-in-1 through 1-in-8
+but not at probabilities 1-in-16 and smaller.
+
+### Proposed `tracestate` syntax
+
+The consistent sampling randomness valuw (`r`) and and head sampling
+probability value (`p`) will be propagated using two bytes of base16 content
+for each of the two fields, as follows:
+
+```
+tracestate: otel=p:PP;r:RR
+```
+
+where `PP` are two bytes of base16 probability value and `RR` are two
+bytes of base16 random value.  These values are omitted when they are
+unknown.
+
+This proposal should be taken as a recommendation and will be modified
+to [match whatever format OpenTelemtry specifies for its
+`tracestate`](https://github.com/open-telemetry/opentelemetry-specification/pull/1852).
+The choice of base16 encoding is therefore just a recommendation,
+chosen because `traceparent` uses base16 encoding.
+
+### Examples
+
+The following `tracestate` value:
+
+```
+tracestate: otel=r:0a;p:03
+```
+
+translates to
+
+```
+base16(probability) = 03 // 1-in-4 head probability
+base16(randomness) = 0a // qualifies for 1-in-1024 or greater probability consistent sampling
+```
+
+Any `TraceIDRatioBased` Sampler configured with probability 2**-9 or
+greater will enable sampling this trace, whereas any
+`TraceIDRatioBased` Sampler configured with probability 2**-10 or less
+will stop sampling this trace.
+
+## Internal details
+
+The reasoning behind restricting the set of sampling rates is that it:
+
+- Lowers the cost of propagating head sampling probability
+- Limits the number of random bits required
+- Avoids floating-point to integer rounding errors
+- Makes math involving partial traces tractable.
+
+[An algorithm for making statistical inference from partially-sampled
+traces has been published](https://arxiv.org/pdf/2107.07703.pdf) that
+explains how to work with a limited number of power-of-2 sampling rates.
+
+### Behavior of the `TraceIDRatioBased` Sampler
```

Comment:
I would recommend introducing a new sampler with a different name and not touching the TraceIDRatioBased at all, because the actual mechanism has nothing to do with "trace-id ratio" anymore.

Reply:
What would you propose that we do with `TraceIDRatioBased`, then? It has TODOs in the specification and the implementations we have rely on unspecified use of TraceID bits. If the best we can do is not touch the `TraceIDRatioBased` logic, then I propose we name it to `IncosistentProbabilityBased` and leave it as deprecated with the existing caveat that it should only be used at the root.

From the perspective of this OTEP, traces rooted in `InconsistentProbabilityBased` samplers would not be able to convey head probability.

From the perspective of OTEP 170, all spans in such a trace would have to record unknown adjusted count. We could _in the future_ resume discussions about a `sampler.adjusted_count` attribute to allow arbitrary sampling probability at the root span while propagating unknown head probability. Consumers could then use `InconsistentProbabilityBased` samplers w/ arbitrary probability at the root of their trace, perform trace assembly, and then perform span-to-metrics over sampled traces.

As for naming the proposed new Sampler, @oertl how do you feel about `ConsistentProbabilityBased`? 

(Actually, do we need all these "Based" strings anyway, could we just talk about "Parent", "TraceIDRatio", "ConsisterntProbability", ...?)

Reply:
> What would you propose that we do with TraceIDRatioBased, then

Deprecated it? Since the spec never said how it needs to be implemented, do existing SDKs even have this sampler today? And if they do, how can it be consistent across langs?

Reply:
OK. Yes the SDKs have implemented this Sampler because among the builtin samplers, it's the only one useful as a root other than always-on and always-off. It's just not specified in a consistent way. That's why I proposed it _could_ be kept as a sampler for non-power-of-two-probabilities, which the legacy TraceIDRatioBased sampler supports, but I think it would be better to replace it and round the input sampling probabilities to powers-of-two. It's fine to choose a new name, my goal is that we end up with just one behavior.

Could we just call this single new sampler the ProbabilitySampler? Could we merge this OTEP and debate names and whether to deprecate the legacy behavior in another PR?

Reply:
I would prefer more special name like "ConsistentSampler" over "ProbabilitySampler", as the letter is pretty common in pre-OTEL SDKs and means "flip a p-weighted coin"